### PR TITLE
This branch implements fine-grained TX conflict resolution, utilizing the conflict-parameter annotations.

### PR DESCRIPTION
**Pull request by dahliamalkhi**
This branch is ready to be merged. After this is merged, getting fine-grained TX conflict resolution will be a small step.

### CODE REVIEW

File: test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 import lombok.Getter;
+import org.corfudb.protocols.wireprotocol.CorfuMsgType;
 import org.corfudb.protocols.wireprotocol.LogData;
 import org.corfudb.protocols.wireprotocol.TokenResponse;
 import org.corfudb.runtime.CorfuRuntime;
 import org.corfudb.runtime.clients.SequencerClient;
+import org.corfudb.runtime.clients.TestRule;
 import org.junit.Before;
 import org.junit.Test;
+import sun.rmi.runtime.Log;
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java]: This is probably the wrong import

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ICorfuPayload.java [hunk 2]
```diff
@@ -124,6 +124,27 @@
         return builder.build();
     }
 
+    /** A TxResolutionInfo implementation.
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/ICorfuPayload.java]: Do we really need a TXResolutionInfo type? It appears that this is just a way to abstract out the combination of the first read timestamp and the conflict map, which could easily just be part of the request as previous, with the timestamp and conflict map just set to null if not used.

Alternatively, you could have two message types: token_req and token_tx_req. Adding TXResolutionInfo as a type the serializer has to deal with seems like a bad idea.

**Comment by dahliamalkhi** [comment]
@no2chem  Most of your comments have already been addressed by the latest commit. Please re-check.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/ReadSetEntry.java [hunk 3]
```diff
@@ -5,6 +5,7 @@
 /** An entry in a transaction read set.
  * Created by mwei on 12/19/16.
  */
+@Deprecated
 public class ReadSetEntry {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/ReadSetEntry.java]: This isn't really a publicly exposed-api, so why should it be marked deprecated? Just remove the class and any references if you're going to deprecate it.

**Comment by dahliamalkhi** [comment]
The fine-grained TX resolution implementation is now completed, and @no2chem 's review comments addressed. @no2chem waiting for you to review and approve.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 4]
```diff
@@ -376,30 +337,20 @@ private OptimisticTransactionalContext getRootContext() {
      *
      * @return The first timestamp to be used for this transaction.
      */
-    private synchronized long fetchFirstTimestamp() {
-        if (getRootContext() != this) {
+    @Override
+    public synchronized long obtainSnapshotTimestamp() {
+        AbstractTransactionalContext atc = getRootContext();
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: should be marked final

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 5]
```diff
@@ -47,23 +48,28 @@
     private final Map<String, Object> opts;
 
     /**
-     * The sequencer maintains information about log and streams:
+     * The sequencer maintains information about log and branches:
```
Comment (no2chem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: did we agree to revert this back to streams?

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 6]
```diff
@@ -209,14 +189,14 @@
...
-        writeSet.putIfAbsent(proxy.getStreamID(), new LinkedList<>());
-        // Insert the modification into our write set.
-        writeSet.get(proxy.getStreamID()).add(
-                new WriteSetEntry(updateEntry, conflictObject));
+                              Object[] conflictObjects) {
+
+        // Insert the modification into writeSet.
+        addToWriteSet(proxy, updateEntry, conflictObjects);
+
+        // TODO: what is this?
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: can you leave this TODO comment out?

Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: done

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 4 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 5 above)
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: ok ok, done.

**Pull request merged**